File: python/102/patterns/shoppingcartcommand.py
# Command Interface
class Command:
    def execute(self):
        pass

    def undo(self):
        pass

# Receiver
class ShoppingCart:
    def __init__(self):
        self.items = []

    def add_item(self, item):
        self.items.append(item)
        print(f"Added {item} to the cart. Current items: {self.items}")

    def remove_item(self, item):
        self.items.remove(item)
        print(f"Removed {item} from the cart. Current items: {self.items}")

# Concrete Commands
class AddItemCommand(Command):
    def __init__(self, cart, item):
        self.cart = cart
        self.item = item

    def execute(self):
        self.cart.add_item(self.item)

    def undo(self):
        self.cart.remove_item(self.item)

class RemoveItemCommand(Command):
    def __init__(self, cart, item):
        self.cart = cart
        self.item = item

    def execute(self):
        self.cart.remove_item(self.item)

    def undo(self):
        self.cart.add_item(self.item)

# Invoker
class CommandManager:
    def __init__(self):
        self.history = []

    def execute_command(self, command):
        command.execute()
        self.history.append(command)

    def undo_last_command(self):
        if self.history:
            last_command = self.history.pop()
            last_command.undo()
        else:
            print("Nothing to undo!")

# Client Code
cart = ShoppingCart()
command_manager = CommandManager()

# Create commands
add_apple = AddItemCommand(cart, "Apple")
add_banana = AddItemCommand(cart, "Banana")
remove_apple = RemoveItemCommand(cart, "Apple")

# Execute commands
command_manager.execute_command(add_apple)   # Add Apple
command_manager.execute_command(add_banana) # Add Banana
command_manager.execute_command(remove_apple) # Remove Apple

# Undo commands
command_manager.undo_last_command()  # Undo remove Apple
command_manager.undo_last_command()  # Undo add Banana
command_manager.undo_last_command()  # Undo add Apple
command_manager.undo_last_command()  # Nothing to undo

File: vanilla-js/oop/shoppingcartcommand.js
// Command Interface
class Command {
    execute() {}
    undo() {}
}

// Receiver
class ShoppingCart {
    constructor() {
        this.items = [];
    }

    addItem(item) {
        this.items.push(item);
        console.log(`Added ${item} to the cart. Current items: [${this.items.join(', ')}]`);
    }

    removeItem(item) {
        const index = this.items.indexOf(item);
        if (index !== -1) {
            this.items.splice(index, 1);
            console.log(`Removed ${item} from the cart. Current items: [${this.items.join(', ')}]`);
        } else {
            console.log(`Item ${item} not found in the cart.`);
        }
    }
}

// Concrete Commands
class AddItemCommand extends Command {
    constructor(cart, item) {
        super();
        this.cart = cart;
        this.item = item;
    }

    execute() {
        this.cart.addItem(this.item);
    }

    undo() {
        this.cart.removeItem(this.item);
    }
}

class RemoveItemCommand extends Command {
    constructor(cart, item) {
        super();
        this.cart = cart;
        this.item = item;
    }

    execute() {
        this.cart.removeItem(this.item);
    }

    undo() {
        this.cart.addItem(this.item);
    }
}

// Invoker
class CommandManager {
    constructor() {
        this.history = [];
    }

    executeCommand(command) {
        command.execute();
        this.history.push(command);
    }

    undoLastCommand() {
        if (this.history.length > 0) {
            const lastCommand = this.history.pop();
            lastCommand.undo();
        } else {
            console.log("Nothing to undo!");
        }
    }
}

// Client Code
const cart = new ShoppingCart();
const commandManager = new CommandManager();

// Create commands
const addApple = new AddItemCommand(cart, "Apple");
const addBanana = new AddItemCommand(cart, "Banana");
const removeApple = new RemoveItemCommand(cart, "Apple");

// Execute commands
commandManager.executeCommand(addApple);    // Add Apple
commandManager.executeCommand(addBanana);  // Add Banana
commandManager.executeCommand(removeApple); // Remove Apple

// Undo commands
commandManager.undoLastCommand(); // Undo remove Apple
commandManager.undoLastCommand(); // Undo add Banana
commandManager.undoLastCommand(); // Undo add Apple
commandManager.undoLastCommand(); // Nothing to undo
